tools/clang/test/HLSL/ShaderOpArith.xml

```diff
@@ -581,6 +581,216 @@
+
+    <InputElements>
+      <InputElement SemanticName="POSITION" Format="R32G32B32_FLOAT" AlignedByteOffset="0" />
+      <InputElement SemanticName="TEXCOORD" Format="R32G32_FLOAT" AlignedByteOffset="12" />
+    </InputElements>
+    <RenderTargets>
+      <RenderTarget Name="RTarget"/>
+    </RenderTargets>
+    <Shader Name="CS" Target="cs_6_7" EntryPoint="CSMain" Text="@PS"/>
+    <Shader Name="VS" Target="vs_6_7" EntryPoint="VSMain" Text="@PS"/>
+    <Shader Name="MS" Target="ms_6_7" EntryPoint="MSMain" Text="@PS"/>
+    <Shader Name="AS" Target="as_6_7" EntryPoint="ASMain" Text="@PS"/>
+    <Shader Name="PS" Target="ps_6_7" EntryPoint="PSMain">
+      <![CDATA[
+        #define OFFSETS 18
+        struct PSInput {
+          float4 position : SV_POSITION;
+          float2 uv : TEXCOORD;
+        };
+
+        Texture2D<float> g_tex : register(t0);
+        RWStructuredBuffer<uint4> g_bufMain : register(u0);
+        RWStructuredBuffer<uint4> g_bufMesh : register(u1);
+        RWStructuredBuffer<uint4> g_bufAmp : register(u2);
+
+        SamplerState g_samp : register(s0);
+        SamplerComparisonState g_sampCmp : register(s1);
+
+        int2 ClampOffset(int2 offset) {
+          if (offset.x < -8 || offset.x > 7)
+            offset.x = (offset.x << 28) >> 28;
+          if (offset.y < -8 || offset.y > 7)
+            offset.y = (offset.y << 28) >> 28;
+          return offset;
+        }
+
+        uint4 DoTests( int2 coord, int2 offset ) {
+#if  __SHADER_TARGET_MAJOR > 6 || (__SHADER_TARGET_MAJOR == 6 && __SHADER_TARGET_MINOR >= 7)
+          int2 coffset = ClampOffset(offset);
+          float cmp = (coord.y + coffset.y)*1000 + coord.x + coffset.x;
+          float2 fcoord = (coord + 0.5)/1000.0;
+          return uint4( g_tex.Sample(g_samp, fcoord, offset),
+                        g_tex.SampleCmp(g_sampCmp, fcoord, cmp, offset),
+                        g_tex.SampleCmpLevel(g_sampCmp, fcoord, cmp, 0, offset),
+                        g_tex.SampleCmpLevelZero(g_sampCmp, fcoord, cmp, offset));
+#else
+          coord += ClampOffset(offset);
+          float cmp = (coord.y)*1000 + coord.x;
+          float2 fcoord = (coord + 0.5)/1000.0;
+          return uint4( g_tex.Sample(g_samp, fcoord),
+                        g_tex.SampleCmp(g_sampCmp, fcoord, cmp),
+                        g_tex.SampleLevel(g_samp, fcoord, 0) == cmp,
+                        g_tex.SampleCmpLevelZero(g_sampCmp, fcoord, cmp));
+#endif
+        }
+
+        uint4 DoMoarTests( int2 coord, int2 offset ) {
+          uint3 lcoord = uint3(coord, 0);
+#if  __SHADER_TARGET_MAJOR > 6 || (__SHADER_TARGET_MAJOR == 6 && __SHADER_TARGET_MINOR >= 7)
+          float2 fcoord = (coord + 0.5)/1000.0;
+          return uint4( g_tex.Load(lcoord, offset),
+                        g_tex.SampleBias(g_samp, fcoord, -1.0, offset),
+                        g_tex.SampleGrad(g_samp, fcoord, 1.0, 1.0, offset),
+                        g_tex.SampleLevel(g_samp, fcoord, 0, offset));
```

Comment:
`g_tex` only has one mip level, so I'm not sure these are adequately testing Bias/Grad/Level.

Reply:
Simpler that way. It's very hard to anticipate exactly where a mipmapped result will land. I didn't suppose that mip levels had much effect here.

I suppose I could retrieve two values, one without an offset and one with and verify that the differences are as expected. I'd prefer to do that in a followup.